### PR TITLE
[Chronicle-246] Google Chronicle Backstory Fixes & Enhancements

### PR DESCRIPTION
**Pull request by crestdatasystems**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
- Add a strategy to handle the retry attempts for 429 and 500 error 
- Added 'gcb-list-rules' command.

## Minimum version of Cortex XSOAR
- [x] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demo
 Recorded Demo 

https://user-images.githubusercontent.com/60967033/122752988-3a0d6c00-d2af-11eb-8c98-3f62ba2cdf41.mp4

### CODE REVIEW

**Comment by adi88d** [comment]
@crestdatasystems please register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA) before we proceed with the review.

**Comment by adi88d** [comment]
@crestdatasystems please register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA) before we proceed with the review.

**Comment by crestdatasystems** [comment]
@adi88d Let us know if there are any changes that need to be done.

File: Packs/GoogleChronicleBackstory/ReleaseNotes/2_0_1.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#### Classifiers
+##### New: Chronicle - Incoming Mapper
+- Maps incoming Chronicle incident fields. (Available from Cortex XSOAR 6.0.0).
+##### New: Google Chronicle Backstory
+-  (Available from Cortex XSOAR 5.0.0).
+
+
+#### Integrations
+##### Chronicle
+- Added simple_backoff_rules dictionary to track retry attempts for 429 and 500 error and if 400 or 404 error occurs for detection of any rule , it will be skipped. Added 'gcb-list-rules' command.
```
Comment (adi88d) [inline Packs/GoogleChronicleBackstory/ReleaseNotes/2_0_1.md:11]:
```suggestion
- Added simple_backoff_rules dictionary to track retry attempts for 429 and 500 error and if 400 or 404 error occurs for detection of any rule.
- Added the ***gcb-list-rules*** command.
```

File: Packs/GoogleChronicleBackstory/ReleaseNotes/2_0_1.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+##### New: list_detections_hr.md
+-  (Available from Cortex XSOAR 5.0.0).
+
+
+#### Scripts
+##### ChronicleAssetEventsForHostnameWidgetScript
+- Updated the Docker image to demisto/python3:3.9.1.15759.
```
Comment (adi88d) [inline Packs/GoogleChronicleBackstory/ReleaseNotes/2_0_1.md:18]:
```suggestion
- Updated the Docker image to demisto/python3:3.9.5.20070.
```

please change this for all the scripts you updated the docker

File: Packs/GoogleChronicleBackstory/IncidentFields/incidentfield-Chronicle_Rule_Name.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
     "required": false,
     "neverSetAsRequired": false,
     "isReadOnly": false,
-    "useAsKpi": false,
+    "useAsKpi": true,
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GoogleChronicleBackstory/IncidentFields/incidentfield-Chronicle_Rule_Name.json:16]: @adi88d  This was a new request from the client, this change enables graphing functionality.

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py [hunk 4]
```diff
@@ -19,8 +19,14 @@
 
 SCOPES = ['https://www.googleapis.com/auth/chronicle-backstory']
 
-BACKSTORY_API_V1_URL = 'https://backstory.googleapis.com/v1'
-BACKSTORY_API_V2_URL = 'https://backstory.googleapis.com/v2'
+BACKSTORY_API_V1_URL = 'https://{}backstory.googleapis.com/v1'
+BACKSTORY_API_V2_URL = 'https://{}backstory.googleapis.com/v2'
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:23]: @adi88d we only created a global variable

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py [hunk 5]
```diff
@@ -1606,48 +1611,81 @@ def convert_events_to_actionable_incidents(events: list) -> list:
...
 
         try:
             _, raw_resp = get_detections(client_obj, rule_id, max_fetch, start_time, end_time, next_page_token,
                                          alert_state)
         except ValueError as e:
-            demisto.info("Error while fetching incidents: " + str(e))
-            pending_rule_or_version_id.insert(0, rule_id)
+            if str(e).endswith('Reattempt will be initiated.'):
+                attempts = simple_backoff_rules.get('attempts', 0)
+                if attempts < 60:
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:1657]: @adi88d We're retrying an API call for 429 and 500 HTTP status errors. In attempts would get increment in each fetch-incident execution, which is 1min. The number of attempts as 60 was decided by us and the customer.

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py [hunk 6]
```diff
@@ -1606,48 +1611,81 @@ def convert_events_to_actionable_incidents(events: list) -> list:
...
         events, detection_to_process = detection_to_process[:max_fetch], detection_to_process[max_fetch:]
     else:
         events = detection_to_process
         detection_to_process = []
 
-    return events, detection_to_process, detection_to_pull, pending_rule_or_version_id
+    return events, detection_to_process, detection_to_pull, pending_rule_or_version_id, simple_backoff_rules
 
 
 def get_max_fetch_detections(client_obj, start_time, end_time, max_fetch, detection_incidents, detection_to_pull,
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:1634]: @adi88d we can add an explanation for this method and while statement method.

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py [hunk 7]
```diff
@@ -2045,6 +2093,99 @@ def get_user_alert_hr_and_ec(client_obj: Client, start_time: str, end_time: str,
...
+    # get list of rules from Chronicle Backstory
+    json_data = validate_response(client_obj, request_url)
+    if live_rule:
+        if live_rule == 'true':
+            list_live_rule = [rule for rule in json_data.get('rules', []) if rule.get('liveRuleEnabled')]
+        else:
+            list_live_rule = [rule for rule in json_data.get('rules', []) if not rule.get('liveRuleEnabled')]
+        json_data = {
+            'rules': list_live_rule
+        }
```
[the remark this answers is not in the document]
Comment (adi88d) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py]: When you updated this block you break the indent of the code -please fix that

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adi88d) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:1657]: Please use a global variable for the attempts number

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adi88d) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:1634]: @crestdatasystems Please do

**Review by adi88d — CHANGES_REQUESTED** [review]
@crestdatasystems
Please fix the code indent in your python file and make sure the test are passed successfully
and update the docker images for the scripts to `3.9.5.21272` don't forget to update the release notes after.

**Comment by crestdatasystems** [comment]
@adi88d we're done with the changes you suggested.

**Review by adi88d — CHANGES_REQUESTED** [review]
@crestdatasystems We want to schedule a demo for your contribution, please share with us your availability for the next week,
Thanks

File: Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py [hunk 8]
```diff
@@ -2045,6 +2109,99 @@ def get_user_alert_hr_and_ec(client_obj: Client, start_time: str, end_time: str,
...
+    :rtype: Tuple[Dict[str, Any], Dict[str, Any]]
+    :return: ec, json_data: Context data and raw response for the fetched rules
+    """
+    page_size = args.get('page_size', 100)
+    validate_page_size(page_size)
+    page_token = args.get('page_token', '')
+    if int(page_size) > 1000:
+        raise ValueError('Page size should be in the range from 1 to 1000.')
+
+    live_rule = args.get('live_rule', '').lower()
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GoogleChronicleBackstory/Integrations/GoogleChronicleBackstory/GoogleChronicleBackstory.py:2153]: @adi88d argToBoolean raises a Value Error if the argument is null or empty as **live_rule** is not a required argument.
So we decided to continue with the existing logic only

**Comment by crestdatasystems** [comment]
@adi88d Recorded Demo 

https://user-images.githubusercontent.com/60967033/122752988-3a0d6c00-d2af-11eb-8c98-3f62ba2cdf41.mp4

**Pull request merged**